File: problem_1/solution.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

type Pair struct {
	a, b int
}

var exists = struct{}{}

type set struct {
	m map[int]struct{}
}

func NewSet() *set {
	s := &set{}
	s.m = make(map[int]struct{})
	return s
}

func (s *set) Add(value int) {
	s.m[value] = exists
}

func (s *set) Contains(value int) bool {
	_, c := s.m[value]
	return c
}

// Method that returns all pair values which sum is equal to given number
func getPairs(numbers []int, sum_val int) []Pair {
	s := NewSet()
	pairs := []Pair{}
	for _, val := range numbers {
		diff := sum_val - val
		if s.Contains(diff) {
			pairs = append(pairs, Pair{diff, val})
		}
		s.Add(val)
	}
	return pairs
}

func main() {
	file, err := os.Open("test_cases.txt")
	if err != nil {
		fmt.Println(err)
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		// Reading text_cases.txt in variables
		content := strings.Split(scanner.Text(), " ")
		str_arr := strings.Split(content[0], ",")
		arr := []int{}
		for _, i := range str_arr {
			j, err := strconv.Atoi(i)
			if err != nil {
				fmt.Println(err)
			}
			arr = append(arr, j)
		}
		k, k_err := strconv.Atoi(content[1])
		if k_err != nil {
			fmt.Println(err)
		}

		pair_vals := getPairs(arr, k)
		if len(pair_vals) > 0 {
			fmt.Printf("Following are the pairs which sum is equal to %v: %v.\n", k, pair_vals)
		} else {
			fmt.Printf("No pair is present which sum is euqal to %v.\n", k)
		}

	}
	if err := scanner.Err(); err != nil {
		fmt.Println(err)
	}

}

File: problem_1/solution.py
"""
Identify 2 integer which sum is equal to given number
"""

def get_pairs(list_values, sum_val):
    """Return a list of tuple pair which sum is equal to given sum_val.

    Parameters
    ----------
    list_values : type
        `list_values` contains all the numbers.
    sum_val : type
        Two numbers from `list_values` should be equal to `sum_val`.

    Returns
    -------
    type
        List of tuple pairs.

    """
    set_values = set()
    ans = list()
    for val in list_values:
        diff = sum_val - val
        if diff in set_values:
            ans.append((diff, val))
        set_values.add(val)
    return ans


if __name__ == '__main__':
    filename = "test_cases.txt"
    with open(filename) as file:
        content = file.readline()
        while content:
            arr, k = content.split()
            k = int(k)
            arr = [int(a) for a in arr.split(',')]

            pairs = get_pairs(arr, k)
            if bool(pairs):
                print("Following are the pairs which sum is equal to {}: {}."
                      .format(k, pairs))
            else:
                print("No pair is present which sum is euqal to {}.".format(k))

            content = file.readline()
